File: Python/data-structures.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def __repr__(self):
        return f"ListNode({self.val}, {self.next})"


# A binary tree node
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val 
        self.left = left 
        self.right = right 
    def __repr__(self):
        return f"TreeNode(val={self.val}, left={self.left}, right={self.right})"
        
# a sample tree
tree = TreeNode(
    val=9, 
    left=TreeNode(
        val=20, 
        left=TreeNode(
            val=23,
            left=TreeNode(
                val=42)), 
        right=TreeNode(
            val=31,)
    ), 
    right=TreeNode(
        val=15,
        right=TreeNode(
            val=18))
    )

#           9
#         /   \
#       20     15
#      /  \      \
#     23  31      18
#    /
#   42

# a binary search tree
tree = TreeNode(
    val=6,
    left=TreeNode(
        val=2, 
        left=TreeNode(0),
        right=TreeNode(
            val=4,
            left=TreeNode(3),
                right=TreeNode(5)
            )
    ),
    right=TreeNode(
        val=8, 
        left=TreeNode(7),
        right=TreeNode(9)
    )
)


# A Node i.e in a graph
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not none else []
        

# A trie node i.e from word search II
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isWord = False
    def addWord(self, word):
        cur = self 
        for c in word:
            if c not in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]
        cur.isWord = True


# a trie's separate class dictnode
class DictNode:
    isWord = False
    children = {}
    
# the actual trie
class Trie:
    def __init__(self):
        self.root = DictNode
    def addWord(self, word):
        curr = self.root 
        for char in word:
            if char not in curr.children:
                curr.children[char] = DictNode()
            curr = curr.children[char]
        curr.isWord = True

File: C++/data-structures.cpp
# include<iostream>
# include <string>
# include <unordered_map>



struct ListNode{
	int val;
	ListNode* next;

	ListNode(int value): val(value), next(nullptr) {}
};

class LinkedList{
public:
	ListNode* head;

	// constructor
	LinkedList(): head(nullptr) {}

	// helper function to insert values
	void insert(int value){
		ListNode* newNode = new ListNode(value);

		if (head == nullptr) {
			head = newNode;
		} else {
			ListNode* current = head;
			while (current->next != nullptr)
				current = current->next;
			current->next = newNode;
		}
	}

	// helper function to display the linked list
	void display() {
		ListNode* current = head;
		while (current != nullptr) {
			std::cout << current->val << " -> ";
			current = current->next;
		}
		std::cout << "nullptr" << std::endl;
	}
};

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Trie {
public:
	struct TrieNode {
		bool isEndOfWord;
		std::unordered_map<char, TrieNode*> children;

		TrieNode() :isEndOfWord(false) {};
	};

	// the root attribute(a pointer to a TrieNode)
	TrieNode* root;

	// the constructor
	Trie(){ root = new TrieNode();}

	// method for inserting words to the trie
	void insert(std::string word) {
		TrieNode* curr = root;
		for (char c: word) {
			if (curr->children.find(c) == curr->children.end()){
				curr->children[c] = new TrieNode();
			}
			curr = curr->children[c];
		}
		curr->isEndOfWord = true;
	}

	// method to search for a word in a trie
	bool search(std::string word) {
		TrieNode* node = searchPrefix(word);
		return (node != nullptr && node->isEndOfWord);
	}

	// method to search if any word in the trie starts with an input string
	bool startsWith(std::string word) {
		TrieNode* node = searchPrefix(word);
		return node != nullptr;
	}

private:
	TrieNode* searchPrefix(std::string word) {
		TrieNode* curr = root;
		for (char c:word) {
			if (curr->children.find(c)==curr->children.end())
				return nullptr;
			curr = curr->children[c];
		}
		return curr;
	};
};

int main() {
	return 0;
}